core/controllers/editor.py

```diff
@@ -933,3 +947,38 @@ class StartedTutorialEventHandler(EditorHandler):
     def post(self):
         """Handles GET requests."""
         user_services.record_user_started_state_editor_tutorial(self.user_id)
+
+
+class EditorAutosaveHandler(ExplorationHandler):
+    """Handles requests from the editor for draft autosave."""
+
+    @require_editor
+    def put(self, exploration_id):
+        """Handles PUT requests for draft updation."""
+        # Raise an Exception if the draft change list fails non-strict
+        # validation.
+        try:
+            change_list = self.payload.get('change_list')
+            version = self.payload.get('version')
+            exp_services.create_or_update_draft(
+                exploration_id, self.user_id, change_list, version,
+                datetime.datetime.utcnow())
+        except utils.ValidationError as e:
+            # If this exception is raised, let the user know that their changes
+            # have been auto discarded.
+            exp_services.discard_draft(exploration_id, self.user_id)
+            raise self.InvalidInputException(e)
+
+        # If the value passed here is False, have the user discard the draft
+        # changes.
```

Comment:
I was getting confused here so I think it's worth adding a comment that we still save the changes to datastore even if the version is not valid, so that they're available for recovery later.